src/libraries/System.Text.Json/gen/src/JsonSourceGenerator.cs

```diff
@@ -0,0 +1,117 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json.SourceGeneration.Reflection;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace System.Text.Json.SourceGeneration
+{
+    /// <summary>
+    /// Generates source code to optimize serialization and deserialization with JsonSerializer.
+    /// </summary>
+    [Generator]
+    public sealed class JsonSourceGenerator : ISourceGenerator
+    {
+        private JsonSourceGeneratorHelper? _helper;
+
+        /// <summary>
+        /// Helper for unit tests.
+        /// TODO: refactor this after adopting emitter/parser pattern.
+        /// </summary>
+        public Dictionary<string, Type>? SerializableTypes => _helper.GetSerializableTypes();
+
+        /// <summary>
+        /// Registers a syntax resolver to receive compilation units.
+        /// </summary>
+        /// <param name="context"></param>
+        public void Initialize(GeneratorInitializationContext context)
+        {
+            context.RegisterForSyntaxNotifications(() => new JsonSerializableSyntaxReceiver());
+        }
+
+        /// <summary>
+        /// Generates source code to optimize serialization and deserialization with JsonSerializer.
+        /// </summary>
+        /// <param name="executionContext"></param>
+        public void Execute(GeneratorExecutionContext executionContext)
+        {
+            Compilation compilation = executionContext.Compilation;
+
+            const string JsonSerializableAttributeName = "System.Text.Json.Serialization.JsonSerializableAttribute";
+            INamedTypeSymbol jsonSerializableAttribute = compilation.GetTypeByMetadataName(JsonSerializableAttributeName);
+            if (jsonSerializableAttribute == null)
+            {
+                return;
+            }
+
+            JsonSerializableSyntaxReceiver receiver = (JsonSerializableSyntaxReceiver)executionContext.SyntaxReceiver;
+            MetadataLoadContextInternal metadataLoadContext = new(compilation);
+
+            TypeExtensions.NullableOfTType = metadataLoadContext.Resolve(typeof(Nullable<>));
+
+            JsonSourceGeneratorHelper helper = new(executionContext, metadataLoadContext);
+            _helper = helper;
+
+            // Discover serializable types indicated by JsonSerializableAttribute.
+            foreach (CompilationUnitSyntax compilationUnit in receiver.CompilationUnits)
+            {
+                SemanticModel compilationSemanticModel = executionContext.Compilation.GetSemanticModel(compilationUnit.SyntaxTree);
+
+                foreach (AttributeListSyntax attributeListSyntax in compilationUnit.AttributeLists)
+                {
+                    AttributeSyntax attributeSyntax = attributeListSyntax.Attributes.Single();
```

Comment:
Is Single() what was intended here, rather than First()?  Just want to confirm.  First will give you back the first one, throwing if one doesn't exist.  Single will give you back the first one, throwing if either one doesn't exist or there are multiple (which means it needs to fully iterate the source after finding one to see if there's another).

Reply:
Thanks, I've cleaned this up. `First()` is what I need here.